lib/src/working_copy.rs

```diff
@@ -511,6 +559,64 @@ impl TreeState {
             })
             .collect();
 
+        let fsmonitor_clock_needs_save = fsmonitor_kind.is_some();
+        let changed_files: Option<Vec<PathBuf>> = match fsmonitor_kind {
+            None => None,
+            Some(FsmonitorKind::Test { changed_files }) => Some(changed_files),
+            #[cfg(feature = "watchman")]
+            Some(FsmonitorKind::Watchman) => match self.query_watchman(self.watchman_clock.clone())
+            {
+                Ok((watchman_clock, changed_files)) => {
+                    // TODO: is it okay to update the clock even if we return `Err` below?
+                    self.watchman_clock = Some(watchman_clock);
+                    changed_files
+                }
+                Err(err) => {
+                    eprintln!("Failed to query filesystem monitor: {}", err);
+                    None
+                }
+            },
+            #[cfg(not(feature = "watchman"))]
+            Some(FsmonitorKind::Watchman) => {
+                unimplemented!(
+                    "Cannot query Watchman because jj was not compiled with the `watchman` feature"
+                );
+            }
+        };
+
+        let fsmonitor_matcher: Box<dyn Matcher> = match changed_files {
+            None => Box::new(EverythingMatcher),
+            Some(changed_files) => {
+                let mut repo_paths = Vec::new();
+                for path in changed_files {
+                    let path = match path.strip_prefix(&self.working_copy_path) {
+                        Ok(path) => path,
+                        Err(_) => continue,
+                    };
+                    let repo_path_components = path
+                        .iter()
+                        .map(|component| -> Result<RepoPathComponent, SnapshotError> {
+                            let component = component.to_str().ok_or_else(|| {
+                                SnapshotError::InvalidUtf8Path {
+                                    path: path.as_os_str().to_owned(),
+                                }
+                            })?;
+                            Ok(RepoPathComponent::from(component))
+                        })
+                        .collect::<Result<_, _>>()?;
+
+                    let repo_path = RepoPath::from_components(repo_path_components);
+                    repo_paths.push(repo_path);
+                }
+
+                let repo_path_set: HashSet<_> = repo_paths.iter().collect();
+                deleted_files.retain(|path| repo_path_set.contains(path));
+
+                Box::new(PrefixMatcher::new(&repo_paths))
+            }
+        };
```

Comment:
Could you extract a function for the code from line 563 to here? It looks like it's quite well separated from the rest of the function.